File: shared/src/commonMain/kotlin/pe/ecolecta/domain/model/Entrega.kt
package pe.ecolecta.domain.model

import pe.ecolecta.domain.EntregaInvalidaException

data class Entrega(
    val id: String,
    val jornadaId: String,
    val proveedorId: String,
    val usuarioId: String,
    val zonaId: String,
    val vehiculoId: String,
    val litros: Double,
    val tachos: Int,
    val observaciones: String?,
    val registradoEn: Long,
    val deviceId: String,
    val loteId: String?,
    val anulada: Boolean,
    val syncState: SyncState,
    val syncError: String?,
    val intentos: Int,
    val updatedAt: Long,
    val litrosServidor: Double? = null,
    val tachosServidor: Double? = null,
    val motivoConflicto: String? = null,
) {
    companion object {
        fun crear(
            id: String,
            jornadaId: String,
            proveedorId: String,
            usuarioId: String,
            zonaId: String,
            vehiculoId: String,
            litros: Double,
            tachos: Int,
            observaciones: String?,
            registradoEn: Long,
            deviceId: String,
            loteId: String?,
        ): Result<Entrega> {
            if (litros <= 0.0) return Result.failure(EntregaInvalidaException.LitrosNoPositivos)
            if (tachos <= 0) return Result.failure(EntregaInvalidaException.TachosInvalidos)
            return Result.success(
                Entrega(
                    id = id,
                    jornadaId = jornadaId,
                    proveedorId = proveedorId,
                    usuarioId = usuarioId,
                    zonaId = zonaId,
                    vehiculoId = vehiculoId,
                    litros = litros,
                    tachos = tachos,
                    observaciones = observaciones?.trim()?.ifBlank { null },
                    registradoEn = registradoEn,
                    deviceId = deviceId,
                    loteId = loteId,
                    anulada = false,
                    syncState = SyncState.PENDING,
                    syncError = null,
                    intentos = 0,
                    updatedAt = registradoEn,
                ),
            )
        }
    }
}

File: web/app/Domain/Entregas/Entrega.php
<?php

namespace App\Domain\Entregas;

use App\Domain\Entregas\Exceptions\EntregaInvalidaException;
use DateTimeImmutable;

final class Entrega
{
    private function __construct(
        public readonly ?int $id,
        public readonly int $jornadaId,
        public readonly int $proveedorId,
        public readonly int $usuarioId,
        public readonly int $zonaId,
        public readonly int $vehiculoId,
        public readonly float $litros,
        public readonly int $tachos,
        public readonly ?string $observaciones,
        public readonly DateTimeImmutable $registradoEn,
        public readonly ?string $loteId,
        public readonly bool $anulada,
    ) {}

    public static function crear(
        int $jornadaId,
        int $proveedorId,
        int $usuarioId,
        int $zonaId,
        int $vehiculoId,
        float $litros,
        int $tachos,
        ?string $observaciones,
        DateTimeImmutable $registradoEn,
        ?string $loteId = null,
    ): self {
        self::validar($litros, $tachos);

        return new self(
            id: null,
            jornadaId: $jornadaId,
            proveedorId: $proveedorId,
            usuarioId: $usuarioId,
            zonaId: $zonaId,
            vehiculoId: $vehiculoId,
            litros: $litros,
            tachos: $tachos,
            observaciones: $observaciones !== null ? trim($observaciones) : null,
            registradoEn: $registradoEn,
            loteId: $loteId,
            anulada: false,
        );
    }

    public static function reconstruir(
        int $id,
        int $jornadaId,
        int $proveedorId,
        int $usuarioId,
        int $zonaId,
        int $vehiculoId,
        float $litros,
        int $tachos,
        ?string $observaciones,
        DateTimeImmutable $registradoEn,
        ?string $loteId,
        bool $anulada,
    ): self {
        return new self($id, $jornadaId, $proveedorId, $usuarioId, $zonaId, $vehiculoId, $litros, $tachos, $observaciones, $registradoEn, $loteId, $anulada);
    }

    private static function validar(float $litros, int $tachos): void
    {
        if ($litros <= 0.0) {
            throw EntregaInvalidaException::litrosNoPositivos();
        }

        if ($tachos <= 0) {
            throw EntregaInvalidaException::tachosInvalidos();
        }
    }
}
